### PR TITLE
mds: emit warning with estimated replay time

### PR DESCRIPTION
**Pull request by vshankar**
It is helpful to get an estimate of how much time it might take to finish replaying the journal.

Fixes: https://tracker.ceph.com/issues/61863

Supersedes #52527 

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by vshankar** [comment]
jenkins test make check

**Comment by vshankar** [comment]
https://pulpito.ceph.com/?branch=wip-vshankar-testing-20240226.124304

**Comment by vshankar** [comment]
https://pulpito.ceph.com/vshankar-2024-02-27_04:05:06-fs-wip-vshankar-testing-20240226.124304-testing-default-smithi/

**Comment by vshankar** [comment]
Reruning tests with https://github.com/ceph/ceph/pull/55909 and https://github.com/ceph/ceph/pull/55908 included. See - https://pulpito.ceph.com/vshankar-2024-03-04_08:26:39-fs-wip-vshankar-testing-20240304.042522-testing-default-smithi/

**Comment by vshankar** [comment]
Test failure - https://pulpito.ceph.com/vshankar-2024-03-04_08:26:39-fs-wip-vshankar-testing-20240304.042522-testing-default-smithi/7580900

**Comment by vshankar** [comment]
(dropping from current batch)

File: src/mds/MDLog.cc [hunk 1]
```diff
@@ -170,7 +170,27 @@ uint64_t MDLog::get_safe_pos() const
   return journaler->get_write_safe_pos(); 
 }
 
+// estiamte the replay completion time based on mdlog journal pointers
```
Comment (batrick) [inline src/mds/MDLog.cc:173]:
```suggestion
// estimate the replay completion time based on mdlog journal pointers
```

File: src/mds/mdstypes.h [hunk 2]
```diff
@@ -1020,4 +1020,22 @@ inline bool operator==(const MDSCacheObjectInfo& l, const MDSCacheObjectInfo& r)
 }
 WRITE_CLASS_ENCODER(MDSCacheObjectInfo)
 
+struct EstimatedReplayTime {
+  EstimatedReplayTime(double percent_complete, std::chrono::seconds estimated_time,
+		      std::chrono::seconds elapsed_time)
+    : percent_complete(percent_complete),
+      estimated_time(estimated_time),
+      elapsed_time(elapsed_time) {
+  }
```
Comment (batrick) [inline src/mds/mdstypes.h:1029]: I assume we have a constructor to satisfy Clang but it's not necessary. When constructing do:

`EstimatedReplayTime{percent, est_time, elapsed_time}` instead.

File: src/mds/mdstypes.h [hunk 3]
```diff
@@ -1020,4 +1020,22 @@ inline bool operator==(const MDSCacheObjectInfo& l, const MDSCacheObjectInfo& r)
...
+
+  double percent_complete;
+  std::chrono::seconds estimated_time;
+  std::chrono::seconds elapsed_time;
+};
+
+inline std::ostream& operator<<(std::ostream& out, const EstimatedReplayTime &estimate) {
+  return out << "replay: " << estimate.percent_complete << "\% complete - elapsed time: "
+	     << estimate.elapsed_time << "s, estimated time remaining " << estimate.estimated_time << "s." ;
+}
```
Comment (batrick) [inline src/mds/mdstypes.h:1039]: Suggest adding a print method, using 92df0fcc649345f438d0d7cc6fab658c02060f07. It's cleaner IMO.

File: src/mds/Beacon.cc [hunk 4]
```diff
@@ -528,6 +529,16 @@ void Beacon::notify_health(MDSRank const *mds)
       }
     }
   }
+  if (mds->is_replay()) {
+    CachedStackStringStream css;
+    auto estimate = mds->mdlog->get_estimated_replay_finish_time();
+    if (estimate.elapsed_time > std::chrono::seconds(30)) {
+      *css << "replay: " << estimate.percent_complete << "\% complete - elapsed time: " << estimate.elapsed_time \
```
Comment (batrick) [inline src/mds/Beacon.cc:536]:
```suggestion
      *css << "replay: " << estimate.percent_complete << "% complete - elapsed time: " << estimate.elapsed_time \
```

?

File: qa/tasks/cephfs/test_failover.py [hunk 5]
```diff
@@ -340,6 +340,52 @@ def test_thrash(self):
...
+                health = self.ceph_cluster.mon_manager.get_mon_health(debug=False, detail=True)
+                codes = [s for s in health['checks']]
+                if 'MDS_ESTIMATED_REPLAY_TIME' in codes:
+                    message = health['checks']['MDS_ESTIMATED_REPLAY_TIME']['detail'][0]['message']
+                    time_duration = float(message.split(" ")[-1][0:-2])
+                    completion_percentage = float(message.split(" ")[2][0:-1])
+                    log.debug(f"MDS_ESTIMATED_REPLAY_TIME is present in health: {message}, duration: {time_duration}, completion_percentage: {completion_percentage}")
+                    if completion_percentage >= 50:
+                        return time_duration
+            time.sleep(1)
```
Comment (batrick) [inline qa/tasks/cephfs/test_failover.py:385]: Unless my eyes deceive me, this sleep is not in the `while` loop?

**Review by batrick — CHANGES_REQUESTED** [review]

**Review by batrick — CHANGES_REQUESTED** [review]
Also, I'd like to see the commit broken into:

* `mds: ...`
* `qa: ...`
* `PendingReleaseNotes,doc: ...`

(we need release note)

**Comment by vshankar** [comment]
> Also, I'd like to see the commit broken into:
> 
> * `mds: ...`
> * `qa: ...`
> * `PendingReleaseNotes,doc: ...`
> 
> (we need release note)

@batrick - This PR is _mostly_ @manishym changes that I took over, so many comments from the original PR probably aren't incorporated. I'll plan to visit this soon.

File: qa/tasks/cephfs/test_failover.py (hunk 5 above)
Comment (vshankar) [reply qa/tasks/cephfs/test_failover.py:385]: Tha's correct - need to one level indented.

**Comment by vshankar** [comment]
jenkins retest this please

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
jenkins test make check

**Comment by batrick** [comment]
jenkins test api

**Comment by batrick** [comment]
jenkins test make check arm64

File: src/mds/MDLog.cc [hunk 6]
```diff
@@ -170,7 +170,37 @@ uint64_t MDLog::get_safe_pos() const
   return journaler->get_write_safe_pos(); 
 }
 
+// estimate the replay completion time based on mdlog journal pointers
+EstimatedReplayTime MDLog::get_estimated_replay_finish_time() {
+  ceph_assert(mds->is_replay());
 
+  EstimatedReplayTime estimated_time(0, std::chrono::seconds::zero(), std::chrono::seconds::zero());
```
Comment (batrick) [inline src/mds/MDLog.cc:177]:
```suggestion
  EstimatedReplayTime estimated_time{0, std::chrono::seconds::zero(), std::chrono::seconds::zero()};
```

causes

```
/home/jenkins-build/build/workspace/ceph-api/src/mds/MDLog.cc:177:23: error: no matching constructor for initialization of 'EstimatedReplayTime'
  EstimatedReplayTime estimated_time(0, std::chrono::seconds::zero(), std::chrono::seconds::zero());
                      ^              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/jenkins-build/build/workspace/ceph-api/src/mds/mdstypes.h:1032:8: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 3 were provided
struct EstimatedReplayTime {
       ^
```

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/mds/MDLog.cc (hunk 6 above)
Comment (vshankar) [reply src/mds/MDLog.cc:177]: Nice catch - fixing.

Comment (vshankar) [reply src/mds/MDLog.cc:177]: hmm. I did compile this change before posting.

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/68968](https://tracker.ceph.com/issues/68968).

**Comment by vshankar** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Main#wip-vshankar-testing-20241219063429-debug

File: src/mds/MDLog.cc [hunk 7]
```diff
@@ -221,7 +221,46 @@ uint64_t MDLog::get_safe_pos() const
...
+	   << ", estimated_time=" << std::chrono::round<std::chrono::seconds>(time)
+	   << dendl;
+
+  estimated_time.percent_complete = percent_complete * 100;
+  estimated_time.elapsed_time = elapsed_time;
+  estimated_time.estimated_time = std::chrono::round<std::chrono::seconds>(time);
+  dout(20) << __func__ << "estimated_time.percent_complete=" << estimated_time.percent_complete
+	   << ", estimated_time.elapsed_time=" << estimated_time.elapsed_time
+	   << ", estimated_time.estimated_time=" << estimated_time.estimated_time
+	   << dendl;
```
Comment (mchangir) [inline src/mds/MDLog.cc:261]: Why are there 2 log lines with almost the same info ?

Comment (vshankar) [reply src/mds/MDLog.cc:261]: For debugging when I was trying to get the regex prefect. I'll remove it in a follow up PR once this change is merge (of course if nothing else stands out).

**Review by mchangir — APPROVED** [review]
LGTM

**Pull request merged**